From 670c2f40685a5c6aaa4a292be91f4326846dd157 Mon Sep 17 00:00:00 2001
From: Til Kaiser <mail@tk154.de>
Date: Fri, 10 Jul 2026 21:33:57 +0200
Subject: [PATCH 2/5] bpf: export dynptr size and read-only helpers

Export __bpf_dynptr_size() and __bpf_dynptr_is_rdonly() with
EXPORT_SYMBOL_GPL so GPL-compatible kernel modules can query a dynptr's
size and read-only state.

Remove the static qualifier from bpf_dynptr_get_type() and declare it in
include/linux/bpf.h so it can be called from other kernel translation
units.

Signed-off-by: Til Kaiser <mail@tk154.de>
---
 include/linux/bpf.h  | 1 +
 kernel/bpf/helpers.c | 4 +++-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/include/linux/bpf.h b/include/linux/bpf.h
index e264c695f31c..a9437596224f 100644
--- a/include/linux/bpf.h
+++ b/include/linux/bpf.h
@@ -1386,6 +1386,7 @@ u32 __bpf_dynptr_size(const struct bpf_dynptr_kern *ptr);
 const void *__bpf_dynptr_data(const struct bpf_dynptr_kern *ptr, u32 len);
 void *__bpf_dynptr_data_rw(const struct bpf_dynptr_kern *ptr, u32 len);
 bool __bpf_dynptr_is_rdonly(const struct bpf_dynptr_kern *ptr);
+enum bpf_dynptr_type bpf_dynptr_get_type(const struct bpf_dynptr_kern *ptr);
 int __bpf_dynptr_write(const struct bpf_dynptr_kern *dst, u32 offset,
 		       void *src, u32 len, u64 flags);
 void *bpf_dynptr_slice_rdwr(const struct bpf_dynptr *p, u32 offset,
diff --git a/kernel/bpf/helpers.c b/kernel/bpf/helpers.c
index 5aed5659822f..d40b1e73dcf1 100644
--- a/kernel/bpf/helpers.c
+++ b/kernel/bpf/helpers.c
@@ -1686,6 +1686,7 @@ bool __bpf_dynptr_is_rdonly(const struct bpf_dynptr_kern *ptr)
 {
 	return ptr->size & DYNPTR_RDONLY_BIT;
 }
+EXPORT_SYMBOL_GPL(__bpf_dynptr_is_rdonly);
 
 void bpf_dynptr_set_rdonly(struct bpf_dynptr_kern *ptr)
 {
@@ -1697,7 +1698,7 @@ static void bpf_dynptr_set_type(struct bpf_dynptr_kern *ptr, enum bpf_dynptr_typ
 	ptr->size |= type << DYNPTR_TYPE_SHIFT;
 }
 
-static enum bpf_dynptr_type bpf_dynptr_get_type(const struct bpf_dynptr_kern *ptr)
+enum bpf_dynptr_type bpf_dynptr_get_type(const struct bpf_dynptr_kern *ptr)
 {
 	return (ptr->size & ~(DYNPTR_RDONLY_BIT)) >> DYNPTR_TYPE_SHIFT;
 }
@@ -1706,6 +1707,7 @@ u32 __bpf_dynptr_size(const struct bpf_dynptr_kern *ptr)
 {
 	return ptr->size & DYNPTR_SIZE_MASK;
 }
+EXPORT_SYMBOL_GPL(__bpf_dynptr_size);
 
 static void bpf_dynptr_set_size(struct bpf_dynptr_kern *ptr, u32 new_size)
 {
-- 
2.55.0

